iceoryx_examples/callbacks_in_c/ice_c_callbacks_with_context_data.c

```diff
@@ -0,0 +1,138 @@
+}
+
+typedef struct
+{
+    struct CounterTopic value;
+    bool isSet;
+} cache_t;
+
+typedef struct
+{
+    cache_t leftCache;
+    cache_t rightCache;
+} CounterService;
+
+void onSampleReceivedCallback(iox_sub_t subscriber, void* contextData)
+{
+    CounterService* self = (CounterService*)contextData;
+    const struct CounterTopic* userPayload;
+    if (iox_sub_take_chunk(subscriber, (const void**)&userPayload) == ChunkReceiveResult_SUCCESS)
+    {
+        iox_service_description_t serviceDescription = iox_sub_get_service_description(subscriber);
+        if (strcmp(serviceDescription.instanceString, "FrontLeft") == 0)
+        {
+            self->leftCache.value = *userPayload;
+            self->leftCache.isSet = true;
+        }
+        else if (strcmp(serviceDescription.instanceString, "FrontRight") == 0)
+        {
+            self->rightCache.value = *userPayload;
+            self->rightCache.isSet = true;
+        }
+        printf("received: %d\n", userPayload->counter);
+        fflush(stdout);
+    }
+
+    if (self->leftCache.isSet && self->rightCache.isSet)
+    {
+        printf("Received samples from FrontLeft and FrontRight. Sum of %d + %d = %d\n",
+               self->leftCache.value.counter,
+               self->rightCache.value.counter,
+               self->leftCache.value.counter + self->rightCache.value.counter);
+        fflush(stdout);
+        self->leftCache.isSet = false;
+        self->rightCache.isSet = false;
+    }
+}
+
+int main()
+{
+    signal(SIGINT, sigHandler);
+    signal(SIGTERM, sigHandler);
+
+    iox_runtime_init("iox-c-callback-subscriber");
+
+    // the listener starts a background thread and the callbacks of the attached events
+    // will be called in this background thread when they are triggered
+    iox_listener_storage_t listenerStorage;
+    iox_listener_t listener = iox_listener_init(&listenerStorage);
+
+    iox_sub_options_t options;
+    iox_sub_options_init(&options);
+    options.historyRequest = 10U;
+    options.queueCapacity = 5U;
+    options.nodeName = "iox-c-callback-subscriber-node";
+    iox_sub_storage_t subscriberLeftStorage, subscriberRightStorage;
+
+    CounterService counterService;
+    counterService.leftCache.isSet = false;
+    counterService.rightCache.isSet = false;
+
+    iox_sub_t subscriberLeft = iox_sub_init(&subscriberLeftStorage, "Radar", "FrontLeft", "Counter", &options);
+    iox_sub_t subscriberRight = iox_sub_init(&subscriberRightStorage, "Radar", "FrontRight", "Counter", &options);
+
+    // attach everything to the listener, from here one the callbacks are called when an event occurs
+    // we attach the pointer to counterService as context data with is then provided as second argument to
+    // the callback which allows us to modify counterService from within the callback
+    iox_listener_attach_subscriber_event_with_context_data(
+        listener, subscriberLeft, SubscriberEvent_DATA_RECEIVED, &onSampleReceivedCallback, &counterService);
```

Comment:
It may be worth emphasizing that the user must ensure the argument provided exists when the callback is called, the API is not able to ensure this. (Could be mentioned here at needs and also stated in the readme).